mirgecom/profiling.py

```diff
@@ -123,66 +76,138 @@ class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+
+        args_tuple = tuple(
+            (arg.size)
+            for arg in actual_args if isinstance(arg, cl.array.Array))
+
+        try:
+            self.kernel_stats[knl][args_tuple]
+        except KeyError:
+            nbytes = 0
+            nops = 0
+            for arg in actual_args:
+                if isinstance(arg, cl.array.Array):
+                    nbytes += arg.size * arg.dtype.itemsize
+                    nops += arg.size
+            res = SingleCallKernelProfile(time=0, flops=nops, bytes_accessed=nbytes,
+                                footprint_bytes=nbytes)
+            self.kernel_stats.setdefault(knl, {})[args_tuple] = res
+
+        if self.logmgr and f"{name}_time" not in self.logmgr.quantity_data:
+            self.logmgr.add_quantity(KernelProfile(self, name))
+
+        self.profile_events.append(ProfileEvent(evt, knl, args_tuple))
+
+        return evt
+
+    def _wait_and_transfer_profile_events(self) -> None:
         # First, wait for completion of all events
         if self.profile_events:
             cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
 
-        global nonloopy_profile_events
-        if nonloopy_profile_events:
-            cl.wait_for_events([npe.cl_event for npe in nonloopy_profile_events])
-
         # Then, collect all events and store them
         for t in self.profile_events:
             program = t.program
+            if hasattr(program, "name"):
+                name = program.name
+            else:
+                name = program.function_name
             r = self._get_kernel_stats(program, t.args_tuple)
             time = t.cl_event.profile.end - t.cl_event.profile.start
 
-            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
-
-            self.profile_results.setdefault(program, []).append(new)
+            new = SingleCallKernelProfile(time, r.flops, r.bytes_accessed,
+                                          r.footprint_bytes)
 
-        for t in nonloopy_profile_events:
-            program = t.knl
-            time = t.cl_event.profile.end - t.cl_event.profile.start
-            new = ProfileResult(time, t.nops, t.nbytes, t.nbytes)
-            self.profile_results.setdefault(program, []).append(new)
+            self.profile_results.setdefault(name, []).append(new)
 
         self.profile_events = []
-        nonloopy_profile_events = []
+
+    def get_profiling_data_for_kernel(self, kernel_name: str) \
+          -> MultiCallKernelProfile:
+        """Return profiling data for kernel `kernel_name`."""
+        self._wait_and_transfer_profile_events()
+
+        if kernel_name not in self.profile_results:
+            return MultiCallKernelProfile(None, None, None, None, None)
+
+        knl_results = self.profile_results[kernel_name]
+
+        num_calls = len(knl_results)
+
+        time = sum([v.time for v in knl_results]) / num_calls / 1e9
+
+        if knl_results[0].flops is not None:
```

Comment:
```suggestion
        # otherwise, the key would not exist in self.profile_results
        assert knl_results

        if knl_results[0].flops is not None:
```

Reply:
I don't understand this suggestion @inducer. Isn't the `assert` here superfluous with the `if kernel_name not in self.profile_results:` in line 170? If not, why not put the assert before the determination of `num_calls`?

Reply:
I added it because there was an access to `knl_results[0]` which isn't generically safe. (It's a pattern that my brain just picks out, I guess.) I suggested adding this so that next time I spot it, I don't have to spend time convincing myself over again that it's safe. I'm all in favor of moving it to a more sensible spot.

Reply:
This change isn't necessary anymore since the code uses StatisticsAccumulator now.